### PR TITLE
Add optional timestamp requirement to test-and-set functionality

### PR DESCRIPTION
**Pull request by vekterli**
@havardpe please review 👀

This adds the ability to atomically predicate a conditional mutation (Put/Update/Remove) on the exact timestamp of the document persisted in the backend (it is expected that this timestamp is retrieved earlier via e.g. visiting).

Timestamp predication has multiple benefits:

* Allows for cheaper evaluation of Document V1 batch update/remove, as it avoids having to reevaluate the full condition selection expression when a mutating batch operation hits the backend. Instead, the batch mutation is predicated on the _timestamp_ of the document that matched during the visiting phase. Evaluation a timestamp condition only requires fetching and comparing document metadata, which is very cheap in comparison.
* Allows for supporting imported fields for batch updates/removes, since selections referencing imported fields work for the visiting phase but not for the mutation _selection_ condition evaluation phase. By removing the selection evaluation entirely from the mutation phase, this is no longer an issue.
* Allows for redesigning reindexing visitors to no longer require bucket-level locking to prevent concurrent mutations to documents. Instead, each operation can be individually timestamp-predicated.

In order to gracefully support rolling upgrades, it's possible to create a TaS condition with _both_ a required timestamp as well as a selection expression. This has the following semantics:

* If the backend supports test-and-set conditions with timestamp requirements, the condition passes _iff_ the currently persisted timestamp exactly matches the timestamp in the condition. The selection condition string is completely _ignored_ in this case.
* If the backend is too old to support timestamp requirements the condition selection string is evaluated as if there was _no_ required timestamp present.

To avoid divergent handling across replicas on different versions, the distributor will "normalize" TaS conditions to fit the lowest common denominator during operation processing. Content nodes communicate support for timestamp requirements as part of the bucket info handshake.

This fallback mechanism should be removed on Vespa 9.

Tagging as "do not merge" until the public API changes in `TestAndSetCondition` have been separately reviewed.

### CODE REVIEW

**Comment by vekterli** [comment]
Removing "do not merge" label as all public API changes are annotated with `@Beta` pending review.

File: storage/src/vespa/storage/distributor/node_supported_features.h [hunk 1]
```diff
@@ -11,12 +11,23 @@ namespace storage::distributor {
...
+    bool unordered_merge_chaining               = false;
+    bool two_phase_remove_location              = false;
     bool no_implicit_indexing_of_active_buckets = false;
-    bool document_condition_probe  = false;
+    bool document_condition_probe               = false;
+    bool timestamps_in_tas_conditions           = false;
 
     bool operator==(const NodeSupportedFeatures& rhs) const noexcept = default;
+
+    [[nodiscard]] NodeSupportedFeatures intersection_of(const NodeSupportedFeatures& rhs) const noexcept {
```
Comment (havardpe) [inline storage/src/vespa/storage/distributor/node_supported_features.h:22]: simply intersect?

**Review by havardpe — DISMISSED** [review]
some very minor suggestions

File: document/src/main/java/com/yahoo/document/TestAndSetCondition.java [hunk 2]
```diff
@@ -62,7 +140,10 @@ public String toString() {
         string.append("condition '");
         string.append(conditionStr);
         string.append("'");
-
+        if (requiredTimestamp != 0) {
+            string.append(", expected_timestamp ");
```
[the remark this answers is not in the document]
Comment (vekterli) [reply document/src/main/java/com/yahoo/document/TestAndSetCondition.java]: Renamed to `required_timestamp`

File: storage/src/tests/distributor/twophaseupdateoperationtest.cpp [hunk 3]
```diff
@@ -966,27 +972,56 @@ TEST_F(TwoPhaseUpdateOperationTest, safe_path_condition_mismatch_fails_with_tas_
...
+
+TEST_F(TwoPhaseUpdateOperationTest, safe_path_timestamp_condition_mismatch_fails_with_tas_error) {
+    do_test_safe_path_condition_mismatch_fails_with_tas_error(
+            TestAndSetCondition(120),
+            "Required test-and-set timestamp did not match persisted timestamp");
+}
+
+TEST_F(TwoPhaseUpdateOperationTest, safe_path_timestamp_and_selection_condition_mismatch_fails_with_tas_error) {
+    do_test_safe_path_condition_mismatch_fails_with_tas_error(
+            TestAndSetCondition(120, "testdoctype1.headerval==120"), // Timestamp has precedence
```
[the remark this answers is not in the document]
Comment (vekterli) [reply storage/src/tests/distributor/twophaseupdateoperationtest.cpp]: Fixed

File: storage/src/tests/distributor/twophaseupdateoperationtest.cpp [hunk 4]
```diff
@@ -966,27 +972,56 @@ TEST_F(TwoPhaseUpdateOperationTest, safe_path_condition_mismatch_fails_with_tas_
...
+TEST_F(TwoPhaseUpdateOperationTest, safe_path_selection_condition_match_sends_puts_with_updated_doc) {
+    do_test_safe_path_condition_match_sends_puts_with_updated_doc(TestAndSetCondition("testdoctype1.headerval==110"));
+}
+
+TEST_F(TwoPhaseUpdateOperationTest, safe_path_timestamp_condition_match_sends_puts_with_updated_doc) {
+    do_test_safe_path_condition_match_sends_puts_with_updated_doc(TestAndSetCondition(110)); // matches newest doc
+}
+
+TEST_F(TwoPhaseUpdateOperationTest, safe_path_timestamp_and_selection_condition_match_sends_puts_with_updated_doc) {
+    do_test_safe_path_condition_match_sends_puts_with_updated_doc(TestAndSetCondition(110, "testdoctype1.headerval==110"));
```
[the remark this answers is not in the document]
Comment (vekterli) [reply storage/src/tests/distributor/twophaseupdateoperationtest.cpp]: Fixed

File: storage/src/tests/distributor/twophaseupdateoperationtest.cpp [hunk 5]
```diff
@@ -1581,6 +1631,75 @@ TEST_F(ThreePhaseUpdateTest, single_full_get_tombstone_sends_puts_with_auto_crea
...
+
+TEST_F(ThreePhaseUpdateTest, single_full_get_with_mismatching_timestamp_condition_is_rejected) {
+    do_test_single_full_get_with_mismatching_condition_is_rejected(
+            TestAndSetCondition(1000), // matches old doc, not new
+            "Required test-and-set timestamp did not match persisted timestamp");
+}
+
+TEST_F(ThreePhaseUpdateTest, single_full_get_with_mismatching_timestamp_and_selection_condition_is_rejected) {
+    do_test_single_full_get_with_mismatching_condition_is_rejected(
+            TestAndSetCondition(1000, "testdoctype1.headerval==1000"),
```
[the remark this answers is not in the document]
Comment (vekterli) [reply storage/src/tests/distributor/twophaseupdateoperationtest.cpp]: Fixed

File: storage/src/tests/persistence/testandsettest.cpp [hunk 6]
```diff
@@ -333,7 +406,29 @@ TEST_F(TestAndSetTest, conditional_get_returns_doc_metadata_on_mismatch) {
     EXPECT_FALSE(reply->is_tombstone());
 }
 
-TEST_F(TestAndSetTest, conditional_get_for_non_existing_document_returns_zero_timestamp) {
+TEST_F(TestAndSetTest, conditional_get_with_selection_and_timestamp_only_checks_timestamp) {
+    const api::Timestamp timestamp = 12345;
+    putTestDocument(false, timestamp); // _selection_ does not match, but timestamp does
+    auto reply = invoke_conditional_get(TestAndSetCondition(timestamp));
```
[the remark this answers is not in the document]
Comment (vekterli) [reply storage/src/tests/persistence/testandsettest.cpp]: Fixed

File: storage/src/vespa/storage/distributor/node_supported_features.h (hunk 1 above)
Comment (vekterli) [reply storage/src/vespa/storage/distributor/node_supported_features.h]: Renamed to `intersect`

**Comment by vekterli** [comment]
@havardpe PTAL

**Review by havardpe — APPROVED** [review]

**Pull request merged**